File: player/ultraPlayer/GargleSoundFx.cpp
#include "GargleSoundFx.h"

GargleSoundFx::GargleSoundFx(void):ASoundFx(SoundFxType::Gargle)
{
	_rate = 0;
	_waveform = WaveformType::Triangle;
}

GargleSoundFx::~GargleSoundFx(void)
{
}

float GargleSoundFx::Rate()
{
	return _rate;
}

void GargleSoundFx::Rate( float val )
{
	_rate = val;
}

WaveformType::Type GargleSoundFx::Waveform()
{
	return _waveform;
}

void GargleSoundFx::Waveform( WaveformType::Type val )
{
	_waveform = val;
}

void GargleSoundFx::FromString( string data )
{
	try
	{
		int paramLength = CommunicationProtocol::Instance()->LongNumericParameterLength();
		int index = CommunicationProtocol::Instance()->HeaderLength();

		WaveformToStringConverter* waveformConverter = new WaveformToStringConverter();
		PhaseToStringConverter* phaseConverter = new PhaseToStringConverter();

		_rate = NumericToStringConverter::Convert<float>(data.substr(index, paramLength));
		index += paramLength;
		_waveform = waveformConverter->Convert(data.substr(index, CommunicationProtocol::Instance()->WaveformParameterLength()));
		index += CommunicationProtocol::Instance()->WaveformParameterLength();
		_phase = phaseConverter->Convert(data.substr(index, paramLength));
		// cleanup
		delete waveformConverter;
		delete phaseConverter;
	}
	catch (exception& e)
	{

	}
}

std::string GargleSoundFx::ToString()
{
	throw NotImplementedException("","GargleSoundFx::ToString");
}

PhaseType::Type GargleSoundFx::Phase()
{
	return _phase;
}

void GargleSoundFx::Phase( PhaseType::Type val )
{
	_phase = val;
}

ASoundFx* GargleSoundFx::MakeCopy()
{
	return (ASoundFx*)(new GargleSoundFx(*this));
}

void GargleSoundFx::CopyFrom( ASoundFx* original )
{
	_rate = ((GargleSoundFx*)original)->Rate();
	_waveform = ((GargleSoundFx*)original)->Waveform();
}

File: controller/Model/SoundFx/GargleSoundFx.cs
using System;
using System.Collections.Generic;
using System.Text;
using UltraPlayerController.Resources.Communication;
using Utilities.Math;
using UltraPlayerController.Model.Utilities;
using Utilities.Serialize;

namespace UltraPlayerController.Model.SoundFx
{
    public class GargleSoundFx : ASoundFx
    {
        #region Fields
        private WaveformType _waveform;

        private MathValue _rate;
        #endregion

        #region Constructors
        public GargleSoundFx() : base(SoundFxType.Gargle)
        {
            _rate = new MathValue();

            _rate.Min = 0; // DSFXGARGLE_RATEHZ_MIN
            _rate.Max = 1000; // DSFXGARGLE_RATEHZ_MAX
            _rate .Default = 20; 
            _rate .Unit = MathUnitType.hz ;
            _rate.FriendlyName = "Rate";
            _rate.Step = 0.1f;

            _waveform = WaveformType.Triangle;

        }
        #endregion

        #region Properties
        public MathValue Rate
        {
            get { return _rate; }
        }

        public WaveformType Waveform
        {
            set
            {
                if (value == WaveformType.Sine)
                {
                    throw new ArgumentException("Cannot accept sine waveform");
                }

                _waveform = value;
            }
            get
            {
                return _waveform;
            }
        }

        #endregion

        #region ASoundFx Members

        public override string ParseToString()
        {
            string result = string.Empty;

            try
            {
                result += base.ParseToString();

                result += ValueSerializer.DoFloat(Protocol.Instance.LongNumericParameterLength, _rate.Precision(), _rate.Value);

                WaveformToStringConverter waveConverter = new WaveformToStringConverter();
                result += waveConverter.Convert(_waveform);

            }
            catch (System.Exception)
            {
                // throw
            }

            return result;
        }

        public override void ParseFromString(string textToParse)
        {
            throw new NotImplementedException();
        }

        #endregion
    }
}
